src/_blog/cryptotls.md

```diff
@@ -0,0 +1,60 @@
+---
+tags:
+- quic
+- go
+- standard library
+- crypto/tls
+title: A QUIC API for crypto/tls
+description:
+date: 2023-08-15
+permalink: "/2023-08-15-quic-crypto-tls/"
+translationKey: ''
+header_image: /quic-crypto-tls.png
+author: Marten Seemann
+---
+
+# A QUIC API for crypto/tls
+
+You might have heard us talk about this before: QUIC has become the most important transport in libp2p. For example, in the IPFS network, QUIC connections account for a whopping 80-90% of the connections. go-libp2p uses [quic-go](https://github.com/quic-go/quic-go), a QUIC implementation written in pure Go. quic-go doesn’t only power QUIC support in go-libp2p, but also in [Caddy](https://caddyserver.com/), a full-featured webserver (for its HTTP/3 support), [Adguard](https://github.com/AdguardTeam/AdGuardHome) (for providing DNS over QUIC), and in the synchronization tool [syncthing](https://github.com/syncthing/syncthing/), and [many other projects](https://github.com/quic-go/quic-go#projects-using-quic-go).
+
+## QUIC and TLS 1.3
+
+QUIC uses TLS 1.3 to secure the connection. There’s no such thing as an unencrypted QUIC connection! However, due to running on top of (the unreliable, unordered) UDP, QUIC's  interactions with the TLS stack is quite different from how a TLS connection on top of TCP would look like. The details are described in [RFC 9001](https://www.rfc-editor.org/rfc/rfc9001.html). When QUIC was standardized, it became a necessity for all TLS stacks across languages to expose new APIs. For the longest time, the Go standard library TLS package (crypto/tls) lacked an API for this purpose. We had no choice but to fork crypto/tls to add the required APIs ourselves.
+
+To complicate matters, the quic-go API aimed to accept a regular `tls.Config` (the config struct used to configure the behavior of a TLS connection, defining, among others, the TLS certificates to use), not a type exposed by our TLS fork. This was to ensure that users of the library didn't have to create separate configs for their TCP/TLS servers and their QUIC servers. This is a pretty common use case, e.g. when running HTTP/3 and HTTP/2 in parallel. To meet this requirement, we had to use the dreaded `unsafe` package to convert between `qtls.Config` and `tls.Config`. And since there was no guarantee that the layout of these structs would remain constant between Go releases (in fact, they did change quite frequently), we had to create a new fork of crypto/tls for every new Go version (every 6 months).
```

Comment:
```suggestion
To complicate matters, the quic-go API aimed to accept a regular `tls.Config` (the struct used to configure the behavior of a TLS connection, defining, among others, the TLS certificates to use), not a type exposed by our TLS fork. This was to ensure that users of the library didn't have to create separate configs for their TCP/TLS servers and their QUIC servers. This is a pretty common use case, e.g. when running HTTP/3 and HTTP/2 in parallel. To meet this requirement, we had to use the dreaded `unsafe` package to convert between `qtls.Config` and `tls.Config`. And since there was no guarantee that the layout of these structs would remain constant between Go releases (in fact, they did change quite frequently), we had to create a new fork of crypto/tls for every new Go version (every 6 months).
```